packages/grpc/src/transform.ts

```diff
@@ -0,0 +1,421 @@
+        const valueProto = addType(valueType);
+
+        // A map's value cannot be another map.
+        if (valueProto[0] === $map) {
+          reportDiagnostic(program, {
+            code: "unsupported-field-type",
+            messageId: "recursive-map",
+            target: valueType,
+          });
+          return ref("<unreachable>");
+        }
+
+        return map(keyProto[1] as "string" | ScalarIntegralName, valueProto);
+      }
+
+      // TODO: there are way more scalars in proto than this. How do we expose those knobs to the API writer?
+      const protoType = {
+        bytes: scalar("bytes"),
+        boolean: scalar("bool"),
+        int32: scalar("int32"),
+        int64: scalar("int64"),
+        uint32: scalar("uint32"),
+        uint64: scalar("uint64"),
+        string: scalar("string"),
+        float32: scalar("float"),
+        float64: scalar("double"),
+      }[name];
+
+      if (!protoType) {
+        reportDiagnostic(program, {
+          code: "unsupported-field-type",
+          messageId: "unknown-intrinsic",
+          format: {
+            name: name,
+          },
+          target: t,
+        });
+        return ref("<unreachable>");
+      }
+
+      return protoType;
+    }
+
+    /**
+     * @param model - the Model to convert
+     * @returns a corresponding message declaration
+     */
+    function toMessage(model: ModelType): ProtoMessageDeclaration {
+      return {
+        kind: "message",
+        name: model.name,
+        declarations: [...model.properties.values()].map(toField),
+      };
+    }
+
+    /**
+     * @param property - the ModelProperty to convert
+     * @returns a corresponding field declaration
+     */
+    function toField(property: ModelTypeProperty): ProtoFieldDeclaration {
+      const field: ProtoFieldDeclaration = {
+        kind: "field",
+        name: property.name,
+        type: addType(property.type),
+        index: program.stateMap(fieldIndexKey).get(property),
+      };
+
+      if (property.type.kind === "Array") field.repeated = true;
+
+      return field;
+    }
+
+    function getEffectiveModelType(
+      model: ModelType,
+      anonymousModelName?: string
+    ): SyntheticModel | undefined {
+      if (effectiveModelTypeCache.has(model)) return effectiveModelTypeCache.get(model);
+
+      const properties = [...model.properties.values()];
+
+      const source = properties[0]?.sourceProperty?.node.parent;
+
+      if (
+        source &&
+        source.kind === SyntaxKind.ModelStatement &&
+        properties.length > 0 &&
+        properties.every((p) => p.sourceProperty?.node.parent === source)
+      ) {
+        // TODO: horrible hack: id.sv?
+        const messageLike: SyntheticModel = {
+          name: (source as any).id.sv,
+          properties,
+          trueModelNode: source,
+        };
+
+        effectiveModelTypeCache.set(model, messageLike);
+
+        visitSynthetic(messageLike);
+        return messageLike;
```

Comment:
@nguerrera this is what I came up with for getting a single spread source if there is one (just by rummaging around in the debugger), but not sure how well this will work in practice. Is there a library function I can use to do this instead or a better way to grab this info?

Reply:
Yes, you should be able to use program.checker.getEffectiveModelType

See https://github.com/microsoft/cadl/pull/506